system-addon/lib/HighlightsFeed.jsm

```diff
@@ -125,6 +120,24 @@ this.HighlightsFeed = class HighlightsFeed {
   }
 
   /**
+   * Compare TopSites and Highlights for urls that might exist in both sections.
+   * If true we need to dedupe Highlights.
+   * @param links new TopSites state
+   */
+  _highlightsWillDedupe(links) {
+    const highlightsIndex = SectionsManager.sections.get(SECTION_ID).order;
+    const rows = this.store.getState().Sections[highlightsIndex].rows;
+    // No data available, we need to fetch and dedupe results.
+    if (!rows.length) {
```

Comment:
This special case seems to do the exact opposite of what the method says it should be doing. If there's no highlights, there's actually nothing to be deduped, i.e., highlights will *not* be deduped.

Reply:
Problem is I kept adding logic to it as new edge cases emerged. Logic should be fine, maybe rename it to something like `shouldFetchHighlights` similar to `shouldUpdate` in React when it receives new props.